File: DSA/patterns/sliding_window/maximumSubarray.js
function maxSubarraySum(arr, k) {
  if (arr.length < k) return null;

  let maxSum = 0;
  let currentSum = 0;

  for (let i = 0; i < k; i++) {
    currentSum += arr[i];
  }
  maxSum = currentSum;

  for (let i = k; i < arr.length; i++) {
    currentSum += arr[i] - arr[i - k];
    maxSum = Math.max(maxSum, currentSum);
  }

  return maxSum;
}

// Example Usage
console.log(maxSubarraySum([2, 3, 5, 2, 8, 1, 5], 3)); // Output: 15

File: DSA/patterns/sliding_window/maximum_subarray.py
#!/bin/python3

'''
    Problem:
    Given an array and a window size k, find the maximum sum of any contiguous subarray of size k.
'''

def max_subarray_sum(arr, k):
    '''
        Solution
    '''
    if len(arr) < k:
        return None

    max_sum = current_sum = sum(arr[:k])  # Initial window sum

    for i in range(k, len(arr)):
        current_sum += arr[i] - arr[i - k]  # Slide the window
        max_sum = max(max_sum, current_sum)

    return max_sum

# Example Usage
print(max_subarray_sum([2, 3, 5, 2, 8, 1, 5], 3))  # Output: 15